src/utils/PSMFeatureExtractor.cpp

```diff
@@ -0,0 +1,304 @@
+#include <OpenMS/config.h>
+
+#include <OpenMS/FORMAT/FileHandler.h>
+#include <OpenMS/FORMAT/FileTypes.h>
+#include <OpenMS/DATASTRUCTURES/StringListUtils.h>
+#include <OpenMS/APPLICATIONS/TOPPBase.h>
+#include <OpenMS/FORMAT/MzIdentMLFile.h>
+#include <OpenMS/FORMAT/IdXMLFile.h>
+#include <OpenMS/CONCEPT/ProgressLogger.h>
+#include <OpenMS/CONCEPT/Constants.h>
+#include <OpenMS/ANALYSIS/ID/PercolatorFeatureSetHelper.h>
+#include <QtCore/QFile>
+#include <QtCore/QDir>
+#include <QtCore/QProcess>
+
+#include <iostream>
+#include <cmath>
+#include <string>
+
+using namespace OpenMS;
+using namespace std;
+
+//-------------------------------------------------------------
+//Doxygen docu
+//-------------------------------------------------------------
+
+/**
+  @page UTILS_PSMFeatureExtractor PSMFeatureExtractor
+
+  @brief PSMFeatureExtractor computes extra features for each input PSM
+
+  @experimental Parts of this tool are still work in progress and usage and input requirements or output might change. (multiple_search_engine, Mascot support)
+
+  <center>
+    <table>
+        <tr>
+            <td ALIGN = "center" BGCOLOR="#EBEBEB"> pot. predecessor tools </td>
+            <td VALIGN="middle" ROWSPAN=2> \f$ \longrightarrow \f$ PSMFeatureExtractor \f$ \longrightarrow \f$</td>
+            <td ALIGN = "center" BGCOLOR="#EBEBEB"> pot. successor tools </td>
+        </tr>
+        <tr>
+            <td VALIGN="middle" ALIGN = "center" ROWSPAN=1> @ref TOPP_PeptideIndexer</td>
+            <td VALIGN="middle" ALIGN = "center" ROWSPAN=1> @ref TOPP_PercolatorAdapter </td>
+        </tr>
+    </table>
+  </center>
+
+  <p>
+PSMFeatureExtractor is search engine sensitive, i.e. it's extra features
+vary, depending on the search engine. Thus, please make sure the input is
+compliant with TOPP SearchengineAdapter output. Also, PeptideIndexer compliant
+target/decoy annotation is mandatory.
+Currently supported search engines are Comet, X!Tandem, MSGF+.
+Mascot support is available but in beta development.
+  </p>
+
+  @note if you have extra features you want to pass to percolator, use the extra
+ flag and list the MetaData entries containing the extra features.
+
+  <B>The command line parameters of this tool are:</B>
+  @verbinclude UTILS_PSMFeatureExtractor.cli
+  <B>INI file documentation of this tool:</B>
+  @htmlinclude UTILS_PSMFeatureExtractor.html
+*/
+
+// We do not want this class to show up in the docu:
+/// @cond TOPPCLASSES
+
+class PSMFeatureExtractor :
+  public TOPPBase
+{
+public:
+  PSMFeatureExtractor() :
+    TOPPBase("PSMFeatureExtractor", "Computes extra features for each input PSM.", false)
+  {
+  }
+
+protected:
+  void registerOptionsAndFlags_()
+  {
+    registerInputFileList_("in", "<files>", StringList(), "Input file(s)", true);
+    setValidFormats_("in", ListUtils::create<String>("mzid,idXML"));
+    registerOutputFile_("out", "<file>", "", "Output file in idXML format", false);
```

Comment:
Please change this to one **mandatory** out parameter that can be either idXML or mzid. Having only optional output files easily leads to problems in KNIME and the user needs to manually inactivate ports